controllers/clustergroupupgrade_controller.go

```diff
@@ -866,119 +899,95 @@ func (r *ClusterGroupUpgradeReconciler) reconcileResources(ctx context.Context,
 }
 
-func (r *ClusterGroupUpgradeReconciler) isClusterCompliantWithPolicy(clusterGroupUpgrade *ranv1alpha1.ClusterGroupUpgrade, clusterName string, policy *unstructured.Unstructured) bool {
-	// Get the status part of the policy.
-	statusObject := policy.Object["status"].(map[string]interface{})
-	// Get the part of the status that holds the clusters' compliance.
-	statusCompliance := statusObject["status"]
-
-	// If the status doesn't contain anything, we don't have enough information to decide, so we assume the
-	// cluster is not compliant with the policy.
-	if statusCompliance == nil {
-		return false
-	}
-
-	subStatus := statusCompliance.([]interface{})
-
-	// Loop through all the NonCompliant clusters in the policy's status.
-	for _, crtSubStatusCrt := range subStatus {
-		crtSubStatusMap := crtSubStatusCrt.(map[string]interface{})
-		// If the cluster is NonCompliant, return false.
-		if crtSubStatusMap["compliant"] == utils.StatusNonCompliant && clusterName == crtSubStatusMap["clustername"].(string) {
-			return false
-		}
-	}
-
-	return true
-}
+/*
+  getClusterComplianceWithPolicy returns the compliance of a certain cluster with a certain policy.
+  based on a policy's status structure which is below.
 
-/* A policy's status has the format below. In this function we are creating a list with the clusternames
-   which are NonCompliant.
-
-status: (map[string]interface{})
-  compliant: NonCompliant
-  placement:
-  - placementBinding: operator-placementbinding
-    placementRule: operator-placementrules
-  status: ([]interface{})
-  - clustername: spoke3
-    clusternamespace: spoke3
-    compliant: NonCompliant
-  - clustername: spoke4
-    clusternamespace: spoke4
+  status:
     compliant: NonCompliant
+    placement:
+    - placementBinding: binding-policy1-common-cluster-version-policy
+      placementRule: placement-policy1-common-cluster-version-policy
+    status:
+    - clustername: spoke1
+      clusternamespace: spoke1
+      compliant: NonCompliant
+    - clustername: spoke4
+      clusternamespace: spoke4
+      compliant: NonCompliant
+
+	returns: *string pointer to a string holding either Compliant/NonCompliant/NotMatchedWithPolicy
+	         error
 */
-func (r *ClusterGroupUpgradeReconciler) getClustersNonCompliantWithPolicy(policy *unstructured.Unstructured) []string {
-	var clustersNonCompliant []string
-	// Get the status part of the policy.
-	statusObject := policy.Object["status"].(map[string]interface{})
-	// Get the part of the status that holds the clusters' compliance.
-	statusCompliance := statusObject["status"]
+func (r *ClusterGroupUpgradeReconciler) getClusterComplianceWithPolicy(
+	clusterName string, policy *unstructured.Unstructured) (*string, error) {
 
-	// If all the clusters are compliant, return nil.
-	if statusCompliance == nil {
-		return nil
+	// Get the compliant status part of the policy.
+	statusObject := policy.Object["status"].(map[string]interface{})
+	if statusObject["compliant"] == nil {
```

Comment:
Under what conditions does the global state go to pending (nil)? Is it possible that the global compliance could be pending but we have a valid status for the cluster we care about? For example what if the inform policy has a cluster which is not part of this upgrade CR that is disconnected from the hub cluster. Would a "pending" state on that one cluster cause the global compliance to be pending as well even though we could still query the substatus for the cluster we are handling?